### PR TITLE
Add java sol to usaco superbull

### PR DESCRIPTION
**Pull request by TruaShamu**

### CODE REVIEW

File: solutions/usaco-531.mdx [hunk 1]
```diff
@@ -121,4 +121,103 @@ int main()
 
 </CPPSection>
 
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <cstdio>
```
Comment (SansPapyrus683) [inline solutions/usaco-531.mdx:128]: wait why this

File: solutions/usaco-531.mdx [hunk 2]
```diff
@@ -121,4 +121,103 @@ int main()
...
+		}
+	}
+	init(n);
+	// sort edges in decreasing order of weight
+	sort(edges.begin(), edges.end(), [](Edge a, Edge b){
+		return a.weight > b.weight;
+	});
+	int eliminated = 0;
+	long long ans = 0;
+	for (Edge e : edges)
```
Comment (SansPapyrus683) [inline solutions/usaco-531.mdx:203]: maybe add a short comment "Run Kruskal's on the edges" or something like that idk

File: solutions/usaco-531.mdx [hunk 3]
```diff
@@ -121,4 +121,103 @@ int main()
...
+			edges.push_back({i, j, ids[i] ^ ids[j]});
+		}
+	}
+	init(n);
+	// sort edges in decreasing order of weight
+	sort(edges.begin(), edges.end(), [](Edge a, Edge b){
+		return a.weight > b.weight;
+	});
+	int eliminated = 0;
+	long long ans = 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-531.mdx:202]: (not necessary) `max_score` or smth? maybe?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
~~also put the dsu stuff in a class lmao~~

File: solutions/usaco-531.mdx [hunk 4]
```diff
@@ -121,4 +121,103 @@ int main()
 
 </CPPSection>
 
+<CPPSection>
```
Comment (maggie-j-liu) [inline solutions/usaco-531.mdx:124]: looks like you put the wrong code? (this is mine 👀 )

File: solutions/usaco-531.mdx (hunk 1 above)
Comment (maggie-j-liu) [reply solutions/usaco-531.mdx:128]: freopen

**Comment by nathangong** [comment]
Is this java code? It seems like youre just adding another C++ code block

**Comment by TruaShamu** [comment]
Sorry, I was dumb and I put the wrong code.
It should be fixed now.

File: solutions/usaco-531.mdx [hunk 5]
```diff
@@ -121,4 +121,59 @@ int main()
 
 </CPPSection>
 
+<JavaSection>
```
Comment (maggie-j-liu) [inline solutions/usaco-531.mdx:124]: this is prim's and the c++ sol is kruskal's, so put this in a separate language section (and maybe add headers to the sections)

File: solutions/usaco-531.mdx [hunk 6]
```diff
@@ -121,4 +121,59 @@ int main()
...
+
+```java
+import java.io.*;
```
Comment (maggie-j-liu) [inline solutions/usaco-531.mdx:127]: don't you need to `import java.util.*` ?

File: solutions/usaco-531.mdx [hunk 7]
```diff
@@ -121,4 +121,59 @@ int main()
...
+
+public class superbull {
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio("superbull");
+		int N = io.nextInt();
+		//Read in the input array.
```
Comment (maggie-j-liu) [inline solutions/usaco-531.mdx:133]:
```suggestion
		// Read in the input array.
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-531.mdx [hunk 8]
```diff
@@ -121,4 +121,59 @@ int main()
 
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.io.*;
+
+public class superbull {
```
Comment (SansPapyrus683) [inline solutions/usaco-531.mdx:129]:
```suggestion
public class Superbull {
```

File: solutions/usaco-531.mdx [hunk 9]
```diff
@@ -121,4 +121,59 @@ int main()
...
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio("superbull");
+		int N = io.nextInt();
+		//Read in the input array.
+		int[] inp = new int[N];
```
Comment (SansPapyrus683) [inline solutions/usaco-531.mdx:134]: better name for `inp` maybe?

File: solutions/usaco-531.mdx [hunk 10]
```diff
@@ -121,4 +121,59 @@ int main()
...
+		for (int i = 0; i < N; i++) {
+			inp[i] = io.nextInt();
+		}
+
+		long ans = 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-531.mdx:139]: (not necessary) maybe `max_cost` instead?

File: solutions/usaco-531.mdx [hunk 11]
```diff
@@ -121,4 +121,59 @@ int main()
...
+			// Pair team1 and team2.
+			ans += score[team2];
+			visited[team2] = true;
+
+			// Re-evaluate all pair scores.
+			for (int i = 0; i < N; i++) {
+				score[i] = Integer.max(score[i], (inp[team2] ^ inp[i]));
+			}
+		}
+
```
Comment (SansPapyrus683) [inline solutions/usaco-531.mdx:166]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Review by darren-yao — APPROVED** [review]

**Comment by TruaShamu** [comment]
Alright, done

**Comment by TruaShamu** [comment]
Ok, I also renamed ans -> max_cost and renamed inp -> id.

File: solutions/usaco-531.mdx [hunk 12]
```diff
@@ -122,3 +122,64 @@ int main()
 </CPPSection>
 
 </LanguageSection>
+
+
+## Prim's Implementation
```
Comment (maggie-j-liu) [inline solutions/usaco-531.mdx:127]: add time complexity

**Review by SansPapyrus683 — APPROVED** [review]
nice
i also just realized where your username came from

File: solutions/usaco-531.mdx [hunk 13]
```diff
@@ -2,7 +2,7 @@
 id: usaco-531
 source: USACO Silver 2015 February
 title: Superbull
-author: Maggie Liu
+author: Maggie Liu, Sofia Yang
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_superbull_silver.html)
```
Comment (ryanchou-dev) [inline solutions/usaco-531.mdx:8]:
```suggestion
[Official Analysis (C++)](http://www.usaco.org/current/data/sol_superbull_silver.html)
```

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
just this :)

File: solutions/usaco-531.mdx [hunk 14]
```diff
@@ -122,3 +122,64 @@ int main()
...
+			max_cost += score[team2];
+			visited[team2] = true;
+
+			// Re-evaluate all pair scores.
+			for (int i = 0; i < N; i++) {
+				score[i] = Integer.max(score[i], (ids[team2] ^ ids[i]));
+			}
+		}
+		pw.println(max_cost);
+		pw.close();
```
Comment (dongliu0426) [inline solutions/usaco-531.mdx:175]: io?

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/usaco-531.mdx (hunk 14 above)
Comment (dongliu0426) [reply solutions/usaco-531.mdx:175]: was this tested?

Comment (TruaShamu) [reply solutions/usaco-531.mdx]: fixed

Comment (TruaShamu) [reply solutions/usaco-531.mdx]: it works
[image](https://ibb.co/wgLHRdn)
[code](https://pastebin.com/fErXW15Y)

File: solutions/usaco-531.mdx [hunk 15]
```diff
@@ -122,3 +122,66 @@ int main()
...
+		for (int i = 0; i < N; i++) {
+			ids[i] = io.nextInt();
+		}
+
+		long max_cost = 0;
+		boolean[] visited = new boolean[N];
+		int[] score = new int[N];
+
+		// Loop through possible choices for the first team.
+		for (int team1 = 0; team1 < N; team1++) {
```
Comment (dongliu0426) [inline solutions/usaco-531.mdx:154]: :/ I don't think prim
> Loop through possible choices for the first team.

Doesn't it loop through N iterations, keeping track of node with the minimum (or maximum in this case) weight? It doesn't have anything to do with `team1`?

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

**Comment by TruaShamu** [comment]
I've changed the variable names, rewrote the comments and also added an explanation to make it seem more intuitive. I hope you're happy and satisfied 😄 

I think it's Prim. If you still think it isn't, what do you think it should be, then?

**Comment by maggie-j-liu** [comment]
> I think it's Prim. If you still think it isn't, what do you think it should be, then?

it is prim's, i think dong was just saying that prim's doesn't loop through possible choices for the first team.

File: solutions/usaco-531.mdx [hunk 16]
```diff
@@ -122,3 +122,85 @@ int main()
 </CPPSection>
 
 </LanguageSection>
+
+
+## Prim's Implementation
+
+	
+### Explanation
```
Comment (maggie-j-liu) [inline solutions/usaco-531.mdx:130]: should be `## Explanation`, move the header and explanation above the Implementation header

File: solutions/usaco-531.mdx [hunk 17]
```diff
@@ -122,3 +122,85 @@ int main()
...
+For each edge, its weight is equal to the XOR of the two node ids.
+
+So, we would like to find the maximum spanning tree in this graph by connecting the edges with the weights.
+
+There are 2 steps in doing so:
+
+
+
+1. Find an unvisited node with the greatest weight, and add it to the MST.
+2. Re-evaluate all the edge weights from the MST  to the other unvisited nodes.
```
Comment (maggie-j-liu) [inline solutions/usaco-531.mdx:143]:
```suggestion
2. Re-evaluate all the edge weights from the MST to the other unvisited nodes.
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Review by maggie-j-liu — APPROVED** [review]

**Review by dongliu0426 — COMMENTED** [review]
looks good, but imo the format would look better if it was something like

```
## Solution 1 - Kruskal

### Explanation

blah blah blah

### Implementation

blah blah blah

## Solution 2 - Prim

### Explanation

blah blah blah

### Implementation

blah blah blah
```

**Review by ryanchou-dev — APPROVED** [review]
probably what @dongliu0426 suggested, but the rest looks good :)

**Comment by TruaShamu** [comment]
done.

**Pull request merged**